code/modules/multiz/movement.dm

```diff
@@ -208,29 +234,112 @@
 	if(locate(/obj/structure/stairs) in landing)
 		return 1
 
+	var/area/area1 = get_area(landing)
+	if(!area1.has_gravity())
+		return 1
+
+	if(istype(src,/mob/living) || isobj(src))
+		if(istype(landing, /turf/simulated/open))
+			z_levels_fallen++ // Don't damage them but keep track of this.
+			return 1
+
+		if(istype(landing, /turf/space))
+			z_levels_fallen = 0 // turns out they didn't hit anything solid. Lucky them.
+			return 1
+
+		if(!z_levels_fallen) // Checks if they only fell down one floor.
+			z_levels_fallen = 1
+
+		var/weight = 1
+		if(ismob(src))
+			var/mob/M = src
+			weight = M.mob_size
+		if(isobj(src))
+			var/obj/O = src
+			weight = O.w_class
+
+
+		var/fallforce = throw_range
+		if(isobj(src))
+			var/obj/O = src
+			fallforce = O.throwforce
+
+		if(weight >= 3 && fallforce <= 5)
+			fallforce = throw_range
+
+		var/speed = z_levels_fallen*throw_speed
+		var/mass = (weight/THROWNOBJ_KNOCKBACK_DIVISOR)+density+opacity
+		var/momentum = speed*mass
+		var/damage = fallforce*(speed/THROWFORCE_SPEED_DIVISOR)*momentum
+
+		var/miss_chance = max(15*(z_levels_fallen-1), 0)
+
+		if(!prob(miss_chance))
+			for(var/mob/living/L in landing)
+				if(L != src)
+
+					if(istype(src,/obj/vehicle))
+						var/obj/vehicle/V = src
+						if(L == V.load)
+							continue
+
+					if(istype(L,/mob/living/carbon/human))
+
+						var/mob/living/carbon/human/H = L
+						H.apply_damage(rand(damage/2, damage), BRUTE, "head")
+						H.apply_damage(damage, BRUTE, "chest")
+
+						if(damage >= THROWNOBJ_KNOCKBACK_SPEED)
+							H.Weaken(rand(damage/4, damage/2))
+					else
+						L.apply_damage(damage, BRUTE)
+						L.apply_damage(rand(damage/2, damage), BRUTE)
+
+					if(damage >= 1)
+						L.visible_message("<span class='warning'>[L] has been hit by [src].</span>")
+						var/attacker = null
+						if(ismob(src))
+							attacker = src
+						admin_attack_log(attacker, L, "fell onto", "was fallen on by", "fell ontop of")
+
+
 	if(istype(landing, /turf/simulated/open))
 		visible_message("\The [src] falls from the level above through \the [landing]!", "You hear a whoosh of displaced air.")
 	else if(!istype(landing, /turf/space))
 		visible_message("\The [src] falls from the level above and slams onto \the [landing]!", "You hear something slam onto the floor.")
 
-/mob/living/handle_fall(var/turf/landing)
+/obj/handle_fall(var/turf/landing)
 	if(..())
 		return 1
 
-	var/area/area1 = get_area(landing)
-	if(!area1.has_gravity())
-		return 1
+	if(!istype(src,/obj/vehicle) && !istype(src,/obj/mecha))
```

Comment:
This really shouldn't be handled here.

It should be put into this clause:
```
	else if (!istype(landing, /turf/space))
		visible_message("\The [src] falls from the level above and slams onto \the [landing]!", "You hear something slam onto the floor.")
		z_levels_fallen = 0
```

And yoinked from everywhere else.